value_source.go

```diff
@@ -0,0 +1,108 @@
+package cli
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+// ValueSource is a source which can be used to look up a value,
+// typically for use with a cli.Flag
+type ValueSource interface {
+	fmt.Stringer
+	fmt.GoStringer
+
+	// Lookup returns the value from the source and if it was found
+	// or returns an empty string and false
+	Lookup() (string, bool)
+}
+
+// ValueSourceChain is a slice of ValueSource that allows for
+// lookup where the first ValueSource to resolve is returned
+type ValueSourceChain struct {
```

Comment:
I would prefer 

```
type ValueSources []ValueSource
```

instead of defining a struct. Of course if we need to store state about sources then struct is needed. Also no need to name it ValueSourceChain

Reply:
I intentionally moved away from a slice type because:
- imho the semantics and behavior of literals and funcs with slice type receivers is **weird**, meaning it's different enough from base types and structs that I'd like to avoid it whenever possible
- wrapping the slice in a struct means that the struct can grow and change without working around the above point

I intentionally changed the name to `ValueSourceChain` because `ValueSources` is only one letter off from `ValueSource` and I wanted there to be more "distance" between these types *at least* for visual differentiation.

WDYT?

Reply:
I think its fine as it is.